data/routers/cohorts-router.js

```diff
@@ -0,0 +1,173 @@
+      try {
+        const addedRecord = await db('cohorts')
+          .where('id', addedCohort[0])
+          .first();
+        res.status(201).json(addedRecord);
+      } catch (err) {
+        res.status(500).json({
+          error: `There was an error while retrieving the added cohort data. ${err}`
+        });
+      }
+    } catch (err) {
+      res.status(500).json({
+        error: `There was an error while adding the cohort data. ${err}`
+      });
+    }
+  }
+});
+
+router.get('/', async (req, res) => {
+  try {
+    const cohorts = await db('cohorts');
+    res.status(200).json(cohorts);
+  } catch (err) {
+    res.status(500).json({
+      error: `There was an error while retrieving the cohorts data. ${err}`
+    });
+  }
+});
+
+router.get('/:id', async (req, res) => {
+  const { id } = req.params;
+  try {
+    const cohort = await db('cohorts')
+      .where({ id })
+      .first();
+    if (cohort) {
+      res.status(200).json(cohort);
+    } else {
+      res.status(404).json({
+        error: 'There is no cohort with the specified ID.'
+      });
+    }
+  } catch (err) {
+    res.status(500).json({
+      error: `There was an error while retrieving the cohort data. ${err}`
+    });
+  }
+});
+
+router.get('/:id/students', async (req, res) => {
+  const id = req.params.id;
+  const cohort = await db('cohorts')
+    .where({ id })
+    .first();
+  if (cohort) {
+    try {
+      const cohortStudents = await db('cohorts')
+        .join('students', 'cohorts.id', 'students.cohort_id')
+        .select('cohorts.name as Cohort', 'students.name as Student Name')
+        .where({ cohort_id: id });
+      res.status(200).json(cohortStudents);
+    } catch (err) {
+      res.status(500).json({
+        error: `There was an error while retrieving the cohort's students data. ${err}`
+      });
+    }
+  } else {
+    res.status(404).json({
+      error: 'There is no cohort with the specified ID.'
+    });
+  }
+});
+
+router.put('/:id', async (req, res) => {
+  const { id } = req.params;
+  const { name } = req.body;
+  if (!name) {
+    res.status(400).json({
+      error: 'Please provide the cohort changes you intend to make.'
+    });
+  } else {
+    try {
+      const updatedCohort = await db('cohorts')
+        .where({ id })
+        .update(req.body);
+      if (updatedCohort) {
+        try {
+          const updatedRecord = await db('cohorts')
+            .where({ id })
+            .first();
+          res.status(200).json(updatedRecord);
+        } catch (err) {
+          res.status(500).json({
+            error: `There was an error while retrieving the updated cohort data. ${err}`
+          });
+        }
+      } else {
+        res.status(404).json({
+          message: 'There is no cohort with the specified ID.'
+        });
+      }
+    } catch (err) {
+      res.status(500).json({
+        error: `There was an error while updating the cohort data. ${err}`
+      });
```

Comment:
this nested try/catch pattern works and is structured well, but it _is_ a lot of extra code -- just make sure you use it only if you want to make customer error messages for each API call in your _try_ blocks